src/Ng/CommandHelpers.cs

```diff
@@ -12,381 +13,170 @@
+                // Remove quotes (if any) from the start and end of the argument value.
+                var argumentValue = args[i + 1].Trim(Arguments.Quote);
+                unprocessedArguments.Add(argumentName, argumentValue);
             }
 
-            var secretInjector = GetSecretInjector(result);
-
-            foreach (string input in inputArgs)
-            {
-                result[input] = secretInjector.InjectAsync(result[input]).Result;
-            }
+            var secretInjector = GetSecretInjector(unprocessedArguments);
 
-            return result;
+            return new SecretDictionary(secretInjector, unprocessedArguments);
         }
 
         private static void TraceRequiredArgument(string name)
         {
-            Console.WriteLine("Required argument \"{0}\" not provided", name);
             Trace.TraceError("Required argument \"{0}\" not provided", name);
         }
 
-        private static bool TryGetArgument(IDictionary<string, string> arguments, string searchArg, out string value, bool required = false)
-        {
-            if (!arguments.TryGetValue(Constants.ArgumentPrefix + searchArg, out value))
-            {
-                if (required)
-                {
-                    TraceRequiredArgument(searchArg);
-                    throw new ArgumentException("Required argument not provided", searchArg);
-                }
-
-                value = null;
-                return false;
-            }
-
-            return true;
-        }
-
-        private static SecretInjector GetSecretInjector(IDictionary<string, string> arguments)
+        private static ISecretInjector GetSecretInjector(IDictionary<string, string> arguments)
         {
             ISecretReader secretReader;
-            string vaultName;
-            if (!TryGetArgument(arguments, Constants.VaultName, out vaultName))
+            try
             {
-                secretReader = new EmptySecretReader();
+                var vaultName = arguments.GetOrNull(Arguments.VaultName);
+                var clientId = arguments[Arguments.ClientId];
+                var certificateThumbprint = arguments[Arguments.CertificateThumbprint];
+                var storeName = arguments.GetOrNull<StoreName>(Arguments.StoreName) ?? StoreName.My;
+                var storeLocation = arguments.GetOrNull<StoreLocation>(Arguments.StoreLocation) ?? StoreLocation.LocalMachine;
+                var shouldValidateCert = arguments.GetOrThrow<bool>(Arguments.ValidateCertificate);
+
+                var keyVaultConfig = new KeyVaultConfiguration(vaultName, clientId, certificateThumbprint, storeName, storeLocation, shouldValidateCert);
+
+                secretReader = new CachingSecretReader(new KeyVaultReader(keyVaultConfig),
+                    arguments.GetOrNull<int>(Arguments.RefreshIntervalSec) ??
+                    CachingSecretReader.DefaultRefreshIntervalSec);
             }
-            else
+            catch (Exception)
```

Comment:
> (Exception) [](start = 17, length = 12)

this hides all exceptions from  the user. If there was an initialization error, I want to know about it. Please leave the previous logic as is.